wasm-sdk/rust/wasm-bpf-binding/src/object.rs

```diff
@@ -0,0 +1,422 @@
+
+            // Add the map to the hashmap
+            obj.maps.insert(map_obj.name().into(), map_obj);
+            map = map_ptr.as_ptr();
+        }
+
+        // Populate obj.progs
+        let mut prog: *mut libbpf_sys::bpf_program = std::ptr::null_mut();
+        loop {
+            // Get the pointer to the next BPF program
+            let prog_ptr = {
+                let next_ptr =
+                    unsafe { libbpf_sys::bpf_object__next_program(obj.ptr.as_ptr(), prog) };
+                match NonNull::new(next_ptr) {
+                    Some(prog_ptr) => prog_ptr,
+                    None => break,
+                }
+            };
+
+            let program = unsafe { Program::new(prog_ptr) }?;
+
+            // Add the program to the hashmap
+            obj.progs.insert(program.name().into(), program);
+            prog = prog_ptr.as_ptr();
+        }
+
+        Ok(obj)
+    }
+
+    /// Get a reference to `Map` with the name `name`, if one exists.
+    pub fn map<T: AsRef<str>>(&self, name: T) -> Option<&Map> {
+        self.maps.get(name.as_ref())
+    }
+
+    /// Get a mutable reference to `Map` with the name `name`, if one exists.
+    pub fn map_mut<T: AsRef<str>>(&mut self, name: T) -> Option<&mut Map> {
+        self.maps.get_mut(name.as_ref())
+    }
+
+    /// Get an iterator over references to all `Map`s.
+    /// Note that this will include automatically generated .data, .rodata, .bss, and
+    /// .kconfig maps. You may wish to filter this.
+    pub fn maps_iter(&self) -> impl Iterator<Item = &Map> {
+        self.maps.values()
+    }
+
+    /// Get an iterator over mutable references to all `Map`s.
+    /// Note that this will include automatically generated .data, .rodata, .bss, and
+    /// .kconfig maps. You may wish to filter this.
+    pub fn maps_iter_mut(&mut self) -> impl Iterator<Item = &mut Map> {
+        self.maps.values_mut()
+    }
+
+    /// Get a reference to `Program` with the name `name`, if one exists.
+    pub fn prog<T: AsRef<str>>(&self, name: T) -> Option<&Program> {
+        self.progs.get(name.as_ref())
+    }
+
+    /// Get a mutable reference to `Program` with the name `name`, if one exists.
+    pub fn prog_mut<T: AsRef<str>>(&mut self, name: T) -> Option<&mut Program> {
+        self.progs.get_mut(name.as_ref())
+    }
+
+    /// Get an iterator over references to all `Program`s.
+    pub fn progs_iter(&self) -> impl Iterator<Item = &Program> {
+        self.progs.values()
+    }
+
+    /// Get an iterator over mutable references to all `Program`s.
+    pub fn progs_iter_mut(&mut self) -> impl Iterator<Item = &mut Program> {
+        self.progs.values_mut()
+    }
+
+    /// Retrieve the underlying [`libbpf_sys::bpf_object`].
+    pub fn as_libbpf_bpf_object_ptr(&self) -> NonNull<libbpf_sys::bpf_object> {
+        self.ptr
+    }
+}
+
+impl Drop for Object {
+    fn drop(&mut self) {
+        unsafe {
+            libbpf_sys::bpf_object__close(self.ptr.as_ptr());
```

Comment:
Why we need a more complete sdk? To provide user the same interface as libbpf-rs, but with inner implementation replaced with wasm-bpf.

So you should use bpf functions provided by wasm-bpf, instead of that from libbpf-sys. The rust binding of the wasm-bpf host functions can be seen at https://github.com/eunomia-bpf/wasm-bpf/tree/main/wasm-sdk/rust/wasm-bpf-binding.

Reply:
Besides, currently you don't need to implement features that was not supported by wasm-bpf.

Reply:
Thank you for your advice~